### PR TITLE
[WIN32SS][FONT] Fix the system logical stock font data

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
JIRA issue: [CORE-14885](https://jira.reactos.org/browse/CORE-14885)
This PR will fix the stock font data to improve font selection.

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Some observations when applying PR-709 as is on top of 0.4.10-dev-370-g8f44035:
-hell does *not* come down to earth, even when applying this on its own right here and right now.
-the fi/fli issue CORE-12091 can not longer be observed with Tahoma 12 in Wordpad (but it's still there, we will have to use smaller font sizes in Wordpad to make it happen now)
-msi dialogs are still too large CORE-13161 and Total Commander fonts CORE-11620 are still too large, which means the freetype.c changes within https://jira.reactos.org/secure/attachment/47727/47727_CORE-11620-totalCommanderFontSizePatch.patch is what actually fixes those.
-the freetype.c changes applied on its own do lead to regressions (in rapps) but 'fix system stock fonts' commit1-changes will mitigate those. *that's what I consider the beneficial part of PR-709 from testers pov*

If others think PR-709 is correct from technical POV (and more like 2k3), then I think it's reasonable to commit it. I did not check that part myself!

**Comment by katahiromz** [comment]
My test program I made:
https://jira.reactos.org/secure/attachment/47775/StockFontsTest.zip

THE TEST RESULTS:
Windows 2003 English:
w2k3-english

Windows 10 Japanese:
w10-japanese

**Comment by katahiromz** [comment]
Windows XP Japanese:
winxp-japaneses

Windows 7 Japanese:
win7-japanese

**Comment by katahiromz** [comment]
Could you try this test program in Russian Windows 2003 or XP?

**Comment by JoachimHenze** [comment]
I have no russian OS available, sorry.

**Comment by katahiromz** [comment]
Thanks.

Win2007 Home Server Russian:
win2007-home-server-russian

**Comment by katahiromz** [comment]
Thanks!

Windows XP x64 SP2 English with integrated Russian MUI:
ada

**Comment by katahiromz** [comment]
It seems like these data are changeable by the user language. Not fixed data.

**Comment by katahiromz** [comment]
ReactOS Japanese after:
ros-japanese-after
Looks good to me.

**Comment by katahiromz** [comment]
ROS Japanese after:
ros-japanese-after
Looks good to me.

ROS Russian after:
ros-russian-after
Looks good to me.

ROS English after:
ros-english-after
Looks good to me.

**Comment by katahiromz** [comment]
@HBelusca: What's your opinion about these screenshots?

**Comment by HBelusca** [comment]
@katahiromz: They look good; I suppose the "discrepancies" between MS and ROS concerning lfPitchAndFamily values come from the fact we do not use the very same font files as Windows?
Also I notice that in ROS Russian the DEVICE_DEFAULT_FONT settings don't completely look the same as SYSTEM_FONT ? (lfCharSet is different).

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -160,6 +160,28 @@ static const CHARSETINFO g_FontTci[MAXTCIINDEX] =
...
 };
 
+BYTE FASTCALL IntCharSetFromCodePage(UINT uCodePage)
+{
+    UINT i;
+
+    if (uCodePage == 1)  // CP_OEMCP
+        return OEM_CHARSET;
+
+    if (uCodePage == 2)  // CP_MACCP
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/freetype.c]: Isn't it possible to use the defines names directly?

File: win32ss/gdi/ntgdi/freetype.c [hunk 2]
```diff
@@ -160,6 +160,28 @@ static const CHARSETINFO g_FontTci[MAXTCIINDEX] =
     { SYMBOL_CHARSET, CP_SYMBOL, {{0,0,0,0},{FS_SYMBOL,0}} }
 };
 
+BYTE FASTCALL IntCharSetFromCodePage(UINT uCodePage)
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/freetype.c]: I would suggest to add a comment that refers to gdi/gdi32/objects/font.c!TranslateCharsetInfo() function, in the  TCI_SRCCODEPAGE case.

Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: What like?

File: win32ss/gdi/ntgdi/freetype.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: We cannot #include <winnls.h> at here, and if included <windows.h> then compilation will be slow...

File: win32ss/gdi/ntgdi/freetype.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/gdi32/objects/font.c [hunk 3]
```diff
@@ -1978,6 +1978,7 @@ TranslateCharsetInfo(
         while (index < MAXTCIINDEX && !(*lpSrc>>index & 0x0001)) index++;
         break;
     case TCI_SRCCODEPAGE:
+        /* ciACP is also referred by IntCharSetFromCodePage() function */
```
Comment (HBelusca) [inline win32ss/gdi/gdi32/objects/font.c]: Actually I was thinking about adding the comment in IntCharSetFromCodePage(), not here.

**Comment by katahiromz** [comment]
English AFTER:
ros-english-after

Japanese AFTER:
ros-japanese-after

Russian AFTER:
ros-russian-after

It's perfect!

File: win32ss/gdi/ntgdi/stockobj.c [hunk 4]
```diff
@@ -140,6 +141,106 @@ IntCreateStockPen(DWORD dwPenStyle,
...
 
+static VOID FASTCALL
+CreateStockFonts(void)
+{
+    USHORT ActiveCodePage, OemCodePage;
+    BYTE bActiveCharSet, bOemCharSet;
+    BOOL bIsCJK;
+    static const WCHAR SimSun[] = { 0x5B8B, 0x4F53, 0 };
+    static const WCHAR MingLiU[] = { 0x7D30, 0x660E, 0x9AD4 };
+    static const WCHAR Batang[] = { 0xBC14, 0xD0D5 };
```
Comment (learn-more) [inline win32ss/gdi/ntgdi/stockobj.c]: Both MingLiU and Batang need a null-terminator, if they are used with wcscpy.

**Comment by HBelusca** [comment]
Can we merge this PR even if PR #713 hasn't been merged yet?

**Comment by katahiromz** [comment]
Yeah, OK.

**Comment by HBelusca** [comment]
Thanks, let's go!

**Pull request merged**